# Exception Handling

- An exception is an abnormal condition that arises in a code sequence at run time. 


## Exception-Handling Fundamentals

- A Java exception is an object that describes an exceptional condition that has occurred in a piece of code. 

- When an exceptional condition arises, an object representing that exception is created and thrown in the method that caused the error. 


- The method may choose to handle the exception itself, or pass it on.

- Either way, at some point, the exception is caught and processed.

- Exceptions can be generated by the Java run-time system, or

- They can be manually generated by your code.


### Java exception handling is managed via five keywords:
- try
- catch
- throw
- throws and finally. 

System generated exceptions are automatically `thrown` by the Java run-time system.
## `try`
- Program statements that you want to monitor for exceptions are contained within a try block. 

## catch
- If an exception occurs within the `try` block, it is thrown.
- Your code can catch this exception (using `catch`) and handle it in some rational manner.

## `throw`
- To manually throw an exception, use the keyword throw.

## `throws`
- Any exception that is thrown out of a method must be specified as such by a throws clause.

## `finally`
- Any code that absolutely must be executed after a try block completes is put in a finally block.


## An exception-handling block:

```java
try {
			
} catch (ArrayIndexOutOfBoundsException e) {
	System.out.println(e);
}catch(Exception e) {
	
}
finally {
	System.out.println("finally block");
}
```

---

## Exception Types

- All exception types are subclasses of the built-in class **`Throwable`**.

- Thus, Throwable is at the top of the exception class hierarchy.

- Immediately below Throwable are two subclasses that partition exceptions into two distinct branches.

- **`Exception`**
- **`Error`**

## The top-level exception hierarchy 

![Exception class hierarchy.](./img/exception-types.png)

## **`Exception`**
- Used for exceptional conditions that user programs should catch. 
- Superclass to create your own custom exception types.

### **`RuntimeException`** 
- An important subclass of Exception RuntimeException.
- Exceptions of this type are automatically defined for the programs that you write and include things such as division by zero and invalid array indexing.


## **`Error`**

- Error are used by the Java run-time system to indicate errors having to do with the run-time environment, itself. 

- Stack overflow is an example of such an error. 

- `Error` defines exceptions that are not expected to be caught under normal circumstances by your program.

--- 

## **Uncaught Exceptions**

### **`The default exception handler`**

- Any exception that is not caught by your program will ultimately be processed by the default handler.

- The default handler displays a string describing the exception, prints a stack trace from the point at which the exception occurred, and terminates the program.

- The default exception handler shows how the entire call stack is displayed:

```java
int d=0;
int a = 42/d;
System.out.println(a);
```
---

## **Using `try and catch`**

- A `try` and its `catch` statement form a unit.

- The scope of the `catch` clause is restricted to those statements specified by the immediately preceding `try` statement.
- A `catch` statement cannot catch an exception thrown by another try statement.

```java
try {
	int d=0;
	int a = 42/d;
	System.out.println("Will never print");
} catch (Exception e) {
	System.out.println("Exception" +e);
}
```
- You cannot use `try` on a single statement.

--- 

## **Multiple `catch` Clauses**

- You can specify two or more catch clauses, each catching a different type of exception.

- Each catch statement is inspected in order, and the first one whose type matches that of the exception is executed.

- After one catch statement executes, the others are bypassed, and execution continues after the `try/catch` block.


```java
    String[] student = {"Maung Maung","Aung Aung"};
    int grade=90;
    try {
    	double avg = grade/0;

    	student[10]="Kyawy Kyaw";

    } catch (ArrayIndexOutOfBoundsException e) {
    	System.out.println(e);
    }catch(Exception e) {
    	System.out.println(e);
    }
```

### **`important to remember :`**

- Exception subclasses must come before any of their superclasses.

-  A `catch` statement that uses a superclass will catch exceptions of that type plus any of its subclasses.

-  Thus, a subclass would never be reached if it came after its superclass.

```java
 try {}
 catch(Exception e) {}
 catch (ArrayIndexOutOfBoundsException e) {}
```
---

## **`Nested try Statements`**

- The try statement can be nested.

- That is, a try statement can be inside the block of another try.

```java
 try {
    //  some code
     try{
         // some code
     }
    catch (ArrayIndexOutOfBoundsException e) {}
 }
 catch(Exception e) {}
```

- Each time a try statement is entered, the context of that exception is pushed on the stack.

- If an inner try statement does not have a catch handler for a particular exception, the stack is unwound and the next try statementâ€™s catch handlers are inspected for a match.

```java
try {
	try {
		int result =100/0;// handle by outer catch
	}catch(ArrayIndexOutOfBoundsException e) {
		// TODO: handle exception
	}
} catch (Exception e) {
	// TODO: handle exception
}
```

- This continues until one of the catch statements succeeds, or until all of the nested try statements are exhausted.

- If no catch statement matches, then the Java run-time system will handle the exception.

- Here is an example that uses nested try statements:


```java
doMath(1);
doMath(1,2);
public static void doMath(int ...args) {
	int a =args.length;
	System.out.println(a);
	try {
		int result = 100/a;
		System.out.printf("Result : %d \n a : %d\n",result,a);
		try {
			if(a==1) a=a/(a-a);
			
			if(a==2) {
				int []inum = {1,2};
				inum[20]=100;
				
			}
		} catch (ArrayIndexOutOfBoundsException e) {
			System.out.println("Array-index-out-of-bound"+e);
		}
	} catch (ArithmeticException e) {
		System.out.println("Divide by zero : " +e);
	}
}
```
---

## **`throw`**

- Your program can throw an exception explicitly, using the `throw` statement. 
- The general form of throw is shown here:
```java
throw ThrowableInstance;
if(book==null){
    throw new ResourceNotFound("There is no book.");
}
```

- ThrowableInstance must be an object of type Throwable or a subclass of Throwable.

-  Primitive types  and non-Throwable classes, such as String and Object, cannot be used as exceptions.
